src/mlpack/methods/reinforcement_learning/async_learning_impl.hpp

```diff
@@ -0,0 +1,108 @@
+/**
+ * @file async_learning_impl.hpp
+ * @author Shangtong Zhang
+ *
+ * This file is the implementation of AsyncLearning class,
+ * which is wrapper for various asynchronous learning algorithms.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_ASYNC_LEARNING_IMPL_HPP
+#define MLPACK_METHODS_RL_ASYNC_LEARNING_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace rl {
+
+template <
+  typename WorkerType,
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType
+>
+AsyncLearning<
+  WorkerType,
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType
+>::AsyncLearning(TrainingConfig config,
+                 NetworkType network,
+                 PolicyType policy,
+                 UpdaterType updater,
+                 EnvironmentType environment):
+  config(std::move(config)),
+  learningNetwork(std::move(network)),
+  policy(std::move(policy)),
+  updater(std::move(updater)),
+  environment(std::move(environment))
+{ /* Nothing to do here. */ };
+
+template <
+  typename WorkerType,
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType
+>
+template <typename Measure>
+void AsyncLearning<
+  WorkerType,
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType
+>::Train(Measure& measure)
+{
+  /**
+   * OpenMP doesn't support shared class member variables.
+   * So we need to copy them to local variables.
+   */
+  NetworkType learningNetwork = std::move(this->learningNetwork);
+  if (learningNetwork.Parameters().is_empty())
+    learningNetwork.ResetParameters();
+  NetworkType targetNetwork = learningNetwork;
+  bool stop = false;
+  size_t totalSteps = 0;
+  PolicyType policy = this->policy;
+
+  // Make sure OpenMP will start desired number of threads.
+  omp_set_dynamic(0);
+  // An extra thread for test.
+  omp_set_num_threads(config.NumOfWorkers() + 1);
```

Comment:
I agree with you that there should be an option to specify the number of worker, but I'm not sure, it's a good idea to create the same amount of threads. It could be faster to create a pool of threads = # of cores, to keep the overhead low. We could test the performance by setting #worker >> #cores and compare the results with both strategies, what do you think?

About the test, we could set omp_set_num_threads(1) before the test starts and reset the value afterwards.